File: patterns/Creational/Builder/Java/RealWorldExample/SQLQueryBuilder.java
package patterns.Creational.Builder.Java.RealWorldExample;


Java
// Builder Design Pattern - SQL Query Builder
//
// This example demonstrates the Factory Method Design Pattern to send various
// types of notifications (Email, SMS).

// The Builder interface declares a set of methods to assemble an SQL query.
interface SQLQueryBuilder {
    SQLQueryBuilder select(String table, String[] fields);
    SQLQueryBuilder where(String field, String value, String operator);
    SQLQueryBuilder limit(int start, int offset);
    String getSQL();
}

// Each Concrete Builder corresponds to a specific SQL dialect and may implement
// the builder steps a little bit differently from the others.
// This Concrete Builder can build SQL queries compatible with MySQL.
class MysqlQueryBuilder implements SQLQueryBuilder {
    protected SQLQuery query;

    public MysqlQueryBuilder() {
        this.query = new SQLQuery();
    }

    protected void reset() {
        this.query = new SQLQuery();
    }

    @Override
    public SQLQueryBuilder select(String table, String[] fields) {
        reset();
        query.base = "SELECT " + String.join(", ", fields) + " FROM " + table;
        query.type = "select";
        return this;
    }

    @Override
    public SQLQueryBuilder where(String field, String value, String operator) {
        if (!"select".equals(query.type) && !"update".equals(query.type) && !"delete".equals(query.type)) {
            throw new RuntimeException("WHERE can only be added to SELECT, UPDATE OR DELETE");
        }
        query.where.add(field + " " + operator + " '" + value + "'");
        return this;
    }

    @Override
    public SQLQueryBuilder limit(int start, int offset) {
        if (!"select".equals(query.type)) {
            throw new RuntimeException("LIMIT can only be added to SELECT");
        }
        query.limit = " LIMIT " + start + ", " + offset;
        return this;
    }

    @Override
    public String getSQL() {
        StringBuilder sql = new StringBuilder(query.base);
        if (!query.where.isEmpty()) {
            sql.append(" WHERE ").append(String.join(" AND ", query.where));
        }
        if (query.limit != null) {
            sql.append(query.limit);
        }
        sql.append(";");
        return sql.toString();
    }
}

// This Concrete Builder is compatible with PostgreSQL. While Postgres is very
// similar to Mysql, it still has several differences. To reuse the common code,
// we extend it from the MySQL builder, while overriding some of the building
// steps.
class PostgresQueryBuilder extends MysqlQueryBuilder {
    @Override
    public SQLQueryBuilder limit(int start, int offset) {
        super.limit(start, offset);
        query.limit = " LIMIT " + start + " OFFSET " + offset;
        return this;
    }
}

// SQLQuery represents the SQL query being built.
class SQLQuery {
    String base;
    String type;
    List<String> where = new ArrayList<>();
    String limit;
}

// Client code to demonstrate the SQL Query Builder.
public class Main {
    public static void main(String[] args) {
        System.out.println("Testing MySQL query builder:");
        clientCode(new MysqlQueryBuilder());
        System.out.println();

        System.out.println("Testing PostgreSQL query builder:");
        clientCode(new PostgresQueryBuilder());
        System.out.println();
    }

    public static void clientCode(SQLQueryBuilder queryBuilder) {
        String query = queryBuilder.select("users", new String[]{"name", "email", "password"})
                .where("age", "18", ">")
                .where("age", "30", "<")
                .limit(10, 20)
                .getSQL();

        System.out.println(query);
    }
}

File: patterns/Creational/Builder/JavaScript/RealWorldExample/SQLQueryBuilder.js
// Builder Design Pattern - SQL Query Builder
//
// This example demonstrates the Factory Method Design Pattern to send various
// types of notifications (Email, SMS).

// SQLQueryBuilder interface declares a set of methods to assemble an SQL query.
class SQLQueryBuilder {
    select(table, fields) {}
    where(field, value, operator = '=') {}
    limit(start, offset) {}
    getSQL() {}
}

// Concrete Builder for MySQL SQL queries.
class MysqlQueryBuilder extends SQLQueryBuilder {
    constructor() {
        super();
        this.reset();
    }

    reset() {
        this.query = {};
    }

    select(table, fields) {
        this.reset();
        this.query.base = `SELECT ${fields.join(', ')} FROM ${table}`;
        this.query.type = 'select';
        return this;
    }

    where(field, value, operator = '=') {
        if (!['select', 'update', 'delete'].includes(this.query.type)) {
            throw new Error('WHERE can only be added to SELECT, UPDATE OR DELETE');
        }
        if (!this.query.where) {
            this.query.where = [];
        }
        this.query.where.push(`${field} ${operator} '${value}'`);
        return this;
    }

    limit(start, offset) {
        if (!['select'].includes(this.query.type)) {
            throw new Error('LIMIT can only be added to SELECT');
        }
        this.query.limit = ` LIMIT ${start}, ${offset}`;
        return this;
    }

    getSQL() {
        let sql = this.query.base;
        if (this.query.where && this.query.where.length > 0) {
            sql += ` WHERE ${this.query.where.join(' AND ')}`;
        }
        if (this.query.limit) {
            sql += this.query.limit;
        }
        return sql + ';';
    }
}

// Concrete Builder for PostgreSQL SQL queries.
class PostgresQueryBuilder extends MysqlQueryBuilder {
    limit(start, offset) {
        super.limit(start, offset);
        this.query.limit = ` LIMIT ${start} OFFSET ${offset}`;
        return this;
    }
}

// Client code to demonstrate the SQL Query Builder.
function clientCode(builder) {
    const query = builder.select('users', ['name', 'email', 'password'])
        .where('age', '18', '>')
        .where('age', '30', '<')
        .limit(10, 20)
        .getSQL();

    console.log(query);
}

console.log('Testing MySQL query builder:');
clientCode(new MysqlQueryBuilder());
console.log();

console.log('Testing PostgreSQL query builder:');
clientCode(new PostgresQueryBuilder());
console.log();
